src/adapters/LayerZeroAdapter.sol

```diff
@@ -0,0 +1,152 @@
+
+    /// @inheritdoc ILayerZeroAdapter
+    function wire(uint16 centrifugeId, uint32 layerZeroEid, address adapter) external auth {
+        sources[layerZeroEid] = LayerZeroSource(centrifugeId, adapter);
+        destinations[centrifugeId] = LayerZeroDestination(layerZeroEid, adapter);
+        emit Wire(centrifugeId, layerZeroEid, adapter);
+    }
+
+    /// @dev Update the LayerZero delegate.
+    function setDelegate(address newDelegate) external auth {
+        endpoint.setDelegate(newDelegate);
+        emit SetDelegate(newDelegate);
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Incoming
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc ILayerZeroReceiver
+    function lzReceive(Origin calldata origin, bytes32, bytes calldata payload, address, bytes calldata)
+        external
+        payable
+    {
+        LayerZeroSource memory source = sources[origin.srcEid];
+        require(source.addr != address(0) && source.addr == origin.sender.toAddressLeftPadded(), InvalidSource());
+        require(msg.sender == address(endpoint), NotLayerZeroEndpoint());
+
+        entrypoint.handle(source.centrifugeId, payload);
+    }
+
+    /// @inheritdoc ILayerZeroReceiver
+    function allowInitializePath(Origin calldata origin) external view override returns (bool) {
+        LayerZeroSource memory source = sources[origin.srcEid];
+        return source.addr != address(0) && source.addr == origin.sender.toAddressLeftPadded();
+    }
+
+    /// @inheritdoc ILayerZeroReceiver
+    function nextNonce(uint32, bytes32) external pure override returns (uint64 nonce) {
+        return 0;
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Outgoing
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IAdapter
+    function send(uint16 centrifugeId, bytes calldata payload, uint256 gasLimit, address refund)
+        external
+        payable
+        returns (bytes32 adapterData)
+    {
+        require(msg.sender == address(entrypoint), NotEntrypoint());
+        LayerZeroDestination memory destination = destinations[centrifugeId];
+        require(destination.layerZeroEid != 0, UnknownChainId());
+
+        MessagingReceipt memory receipt =
+            endpoint.send{value: msg.value}(_params(destination, payload, gasLimit), refund);
+        adapterData = receipt.guid;
+    }
+
+    /// @inheritdoc IAdapter
+    function estimate(uint16 centrifugeId, bytes calldata payload, uint256 gasLimit) external view returns (uint256) {
+        LayerZeroDestination memory destination = destinations[centrifugeId];
+        MessagingFee memory fee = endpoint.quote(_params(destination, payload, gasLimit), address(this));
```

Comment:
Could the estimation fail because `address(this)` is not payable?

Maybe we should pass the `refund` address to this method too, to correctly estimate everything. Refunding will add some cost

Reply:
The last param here is not the refund address, but the sender. See https://github.com/centrifuge/protocol-v3/blob/layerzero-adapter/src/adapters/interfaces/ILayerZeroAdapter.sol#L43